### PR TITLE
improve json response

### PR DESCRIPTION
**Pull request by matsumana**
ObjectMapper is reusable.

# jmh benchmark result

## before

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-0_before-txt)

```
Benchmark                               Mode  Cnt  Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10  0.339 ± 0.017  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  0.685 ± 0.022  ops/s
```

## after

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-1_after-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   5.784 ± 2.484  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  62.863 ± 8.882  ops/s
```

# FYI

- [Presentation: Jackson Performance](https://github.com/FasterXML/jackson-docs/wiki/Presentation:-Jackson-Performance)
- [Jackson Best Practices: Performance](http://wiki.fasterxml.com/JacksonBestPracticesPerformance)

### CODE REVIEW

File: avans/src/main/java/me/geso/avans/jackson/JacksonJsonView.java [hunk 1]
```diff
@@ -11,9 +11,17 @@
...
 public interface JacksonJsonView extends Controller, JSONRendererProvider {
 
+	ObjectMapper mapper = new ObjectMapper();
+
 	@Override
 	public default WebResponse renderJSON(final int statusCode, final Object obj) {
-		final ObjectMapper mapper = this.createObjectMapper();
+
+		// Don't initialized yet?
+		if (mapper.getFactory().getCharacterEscapes() == null) {
```
Comment (moznion) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonView.java]: Does here means "initialize the mapper lazily"?
I thought a way to initialize the mapper at once as following;
- Make `createObjectMapper()` to be static (or extract outer factory, perhaps this is better?)
- Initialize the mapper by that factory method on L14.

How is it?

File: avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java [hunk 2]
```diff
@@ -14,14 +14,15 @@
 
 public interface JacksonJsonParamReader extends JSONParamReader {
 
+	ObjectMapper mapper = new ObjectMapper();
```
Comment (moznion) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: I'm a little bit afraid to expose this mapper as public... But I think no idea, this is a specification restriction of interface.

Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: How about [ObjectReader](https://fasterxml.github.io/jackson-databind/javadoc/2.5/com/fasterxml/jackson/databind/ObjectReader.html)?

```java
    private static final ObjectReader reader =
            new ObjectMapper()
                    .configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false)
                    .reader();
```

In my understand, ObjectReader is immutable if ObjectMapper reference is not expose.

Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: いや、ダメか。もう使ってる可能性あるのか。。。スミマセン

いや違う、今回追加されたから OK なはず。

Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: There are interface private static field sample. But a little dirty.
Maybe adding javadoc for field is enough.

```java
public interface InterfaceContainsPrivateStaticFields {
    class _PrivateStaticFields {
        private static ObjectMapper mapper = new ObjectMapper(); // only accessible in this interface.
    }

    default ObjectReader test() {
        return _PrivateStaticFields.mapper.reader();
    }
}
```

Comment (moznion) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: Or think simple: rename `mapper` to `_mapper`?

Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: ... Yes. It's best! > rename to `_mapper`

**Comment by matsumana** [comment]
Thank you for the comments.

Would you please review again?

# jmh benchmark result

## before `cbed00d`

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-2_before_cbed00d-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   9.180 ± 0.777  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  70.132 ± 1.081  ops/s
```

## after `cbed00d`

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-3_after_cbed00d-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   9.122 ± 0.880  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  69.631 ± 1.548  ops/s
```

**Comment by moznion** [comment]
Looks good!

**Comment by kazuki-ma** [comment]
LGTM.

File: avans/src/main/java/me/geso/avans/jackson/JacksonJsonView.java [hunk 3]
```diff
@@ -11,12 +11,23 @@
 
 public interface JacksonJsonView extends Controller, JSONRendererProvider {
 
+	class _PrivateStaticFields {
+		// only accessible in this interface.
+		private static ObjectMapper _mapper = createObjectMapper();
```
Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonView.java]: `ObjectWriter` is also exist.
(But ok to use ObjectMapper if it private)

File: avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java [hunk 4]
```diff
@@ -37,7 +47,9 @@ public default Object readJsonParam(InputStream is, Class<?> valueType)
 		final byte[] byteArray = output.toByteArray();
 
 		try {
-			return mapper.readValue(byteArray, valueType);
+			ObjectReader reader = _PrivateStaticFields._readerMap
+					.computeIfAbsent(valueType, key -> _PrivateStaticFields._reader.withType(key));
```
Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: Concurrency problem.
Using `ConcurrentHashMap` may be needed.

**Comment by matsumana** [comment]
Would you please review again?

# jmh benchmark result

## before `ff01f8a`

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-4_before_ff01f8a-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   8.462 ± 0.903  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  67.365 ± 4.793  ops/s
```

## after `ff01f8a`

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-5_after_ff01f8a-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   8.899 ± 1.167  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  69.319 ± 1.224  ops/s
```

File: avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java (hunk 4 above)
Comment (kazuki-ma) [inline avans/src/main/java/me/geso/avans/jackson/JacksonJsonParamReader.java]: Ok now.

**Comment by matsumana** [comment]
I forgot about backwards compatibility.
I want to add ObjectWriter/ObjectReader setter methods, to customize ObjectMapper's config.

**Comment by matsumana** [comment]
I modified for backwards compatibility.

Would you please review again?



# jmh benchmark result

## before `2a791ca`

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-6_before_2a791ca-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   2.276 ± 0.580  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  69.266 ± 4.158  ops/s
```

## after `2a791ca`

[log file](https://gist.github.com/matsumana/f2a55485a70170112decb56a63ddd6b7#file-7_after_2a791ca-txt)

```
Benchmark                               Mode  Cnt   Score   Error  Units
JacksonJsonParamReaderBenchmark.bench  thrpt   10   5.112 ± 4.325  ops/s
JacksonJsonViewBenchmark.bench         thrpt   10  69.724 ± 3.048  ops/s
```

**Pull request merged**